subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java

```diff
@@ -0,0 +1,227 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.steps;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSortedMap;
+import com.google.common.collect.Maps;
+import org.gradle.api.UncheckedIOException;
+import org.gradle.internal.classloader.ClassLoaderHierarchyHasher;
+import org.gradle.internal.execution.AfterPreviousExecutionContext;
+import org.gradle.internal.execution.BeforeExecutionContext;
+import org.gradle.internal.execution.CachingResult;
+import org.gradle.internal.execution.Step;
+import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.execution.history.AfterPreviousExecutionState;
+import org.gradle.internal.execution.history.BeforeExecutionState;
+import org.gradle.internal.execution.history.impl.DefaultBeforeExecutionState;
+import org.gradle.internal.execution.impl.OutputFilterUtil;
+import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
+import org.gradle.internal.fingerprint.FileCollectionFingerprint;
+import org.gradle.internal.fingerprint.impl.AbsolutePathFingerprintingStrategy;
+import org.gradle.internal.fingerprint.impl.DefaultCurrentFileCollectionFingerprint;
+import org.gradle.internal.snapshot.FileSystemSnapshot;
+import org.gradle.internal.snapshot.ValueSnapshot;
+import org.gradle.internal.snapshot.ValueSnapshotter;
+import org.gradle.internal.snapshot.impl.ImplementationSnapshot;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Optional;
+
+public class CaptureStateBeforeExecutionStep implements Step<AfterPreviousExecutionContext, CachingResult> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CaptureStateBeforeExecutionStep.class);
+
+    private final ClassLoaderHierarchyHasher classLoaderHierarchyHasher;
+    private final ValueSnapshotter valueSnapshotter;
+    private final Step<? super BeforeExecutionContext, ? extends CachingResult> delegate;
+
+    public CaptureStateBeforeExecutionStep(
+        ClassLoaderHierarchyHasher classLoaderHierarchyHasher,
+        ValueSnapshotter valueSnapshotter,
+        Step<? super BeforeExecutionContext, ? extends CachingResult> delegate
+    ) {
+        this.classLoaderHierarchyHasher = classLoaderHierarchyHasher;
+        this.valueSnapshotter = valueSnapshotter;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public CachingResult execute(AfterPreviousExecutionContext context) {
+        BeforeExecutionState beforeExecutionState = context.getWork().isTaskHistoryMaintained()
```

Comment:
Why don't we do:
```
        Optional<BeforeExecutionState> beforeExecutionState = context.getWork().isTaskHistoryMaintained()
            ? Optional.of(createExecutionState(context))
            : Optional.empty();
```

Reply:
IntelliJ complains about fields or local variables having `Optional` as a type, and it makes some sense to me that it does, though not always. I'm in two minds about this a bit tbh. But in this case I prefer using `null`.

Another option would be to replace `isTaskHistoryMaintained()` with something like making `UnitOfWork.getExecutionHistoryStore()` return an empty `Optional` when history shouldn't be maintained, so we could use `Optional.map()` here.

Reply:
Thinking about this a bit more I think I'd actually like to make `getExecutionHistoryStore()` into an `Optional`, but that requires more changes, so I'd do it separately.

Reply:
Fine with me.